### PR TITLE
Add deriving code for DecEq, and a testing framework

### PR DESCRIPTION
**Pull request by redfish64**
I wanted to make automatic deriving code for DecEq, so I modified your code to do it. I also made a few other changes listed below (which is copy pasted from the commit comment). Let me know what you think. If you don't like it, I don't mind if you don't add it, just thought it might be useful for you.

```
Eq.idr

in eqCon, made it uses a single list rather than two identically lengthed lists for clarity


Ord.idr

wasn't working so I commented out tests to get it to compile

---

DecEq.idr

made new derivor for decidably equality

Note that the nice thing about this is that is proving to the
typechecker that the data type is decidably equal, so if it compiles,
it should work (as long as it doesn't hit bugs in the Idris2
typechecker itself)

---

Util.idr

commented out wrapShow to avoid hole

---

Syntax.idr

added "iAs" which is used by DecEq

---

Types.idr

When determining if isIndex is true, it has a check where it considers all types to
not be indexes. This breaks DecEq, so I removed the check.

Not sure why thats there. Eq and Show (according to the tests) work fine either way.
Not sure about the others.

---

tests/...

Mostly copy pasted Idris2's own testing framework for use here. It is
basically all just golden tests, where it runs the idris2 repl and
executes :printdef for all definitions (eqImplFoo1, eqImplFoo1Fun, ...)

Each test directory has the following:

./run -- shell script that invokes idris2 the proper way
./input -- what is fed into the shell script "run"
./output -- what ./run produced on the last successful test

The Main.idr code will run each test and compare to the previous result,
prompting for replacement of output if there is a difference.

You can try it out by simply installing elab-deriving and going into
tests and running "make"

You can also run tests individually by going into the appropriate testdir
and running:

./run idris2

I moved the ShowTest.idr and EqTest.idr here, as well as created a test for
DecEq.
```

### CODE REVIEW

**Comment by MarcelineVQ** [comment]
Very cool!
eq using two lists was just a holdover from earlier iterations where traversing over things multiple times caused Elab to struggle, Elab is quite a bit faster now.
The check for `not (isType argTy)` is just because it's faster to check that first and exit early instead of running indexOf if we don't have to. I should comment that section since it's pretty unclear what it's doing at a glance.

**Comment by MarcelineVQ** [comment]
Although I see now that your decsEq doesn't work with `not (isType argTy)` present, so I've made a wrong assumption.

**Comment by redfish64** [comment]
Thanks! I'm by no means an expert in Idris, so feel free to correct any questionable code or decisions I made (or eliminate entirely as needed).
